src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs

```diff
@@ -13,26 +16,111 @@ public static string[] GetPortNames()
+                // Typical DosDevice serial port name 'USB#VID_1B4F&PID_214F&MI_00#6&381731fd&0&0000#{86e0d1e0-8089-11d0-9ce4-08003e301f73}'
+                //
+                foreach (string dosName in QueryDosDeviceComPorts(GuidDevInterfaceComPort))
+                {
+                    resultPortNames.Add(dosName);
+                }
+            }
+            else
+            {
+                // Query registry register COM[x] ports
+                //
+                using (RegistryKey serialKey = Registry.LocalMachine.OpenSubKey(@"HARDWARE\DEVICEMAP\SERIALCOMM"))
+                {
+                    if (serialKey != null)
+                    {
+                        foreach (string valueName in serialKey.GetValueNames())
+                        {
+                            string portName = ((string)serialKey.GetValue(valueName));
+
+                            // Add the key's value (portname) if it is valid (Win10IoT places garbage in this Reg Key)
+                            // Filters corrupt chars for COM[x] devices in Registry on Win10IoT
+                            if (portName.StartsWith("COM", StringComparison.InvariantCultureIgnoreCase))
+                            {
+                                resultPortNames.Add(portName);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return resultPortNames.ToArray();
+        }
+
+        private static string[] QueryDosDeviceComPorts(string filterGuid)
+        {
+            // Build a list of all system Com Port device names.
+            // memBuff starts with a small arbitary size and dynamically expands until there is enough room for data returned from Kernal32.QueryDosDevice().
+            uint returnSize = 0;
+            int maxSize = 1024;
+            string allDevices = null;
+            const uint ERROR_INSUFFICIENT_BUFFER = 122;
 
-            using (RegistryKey serialKey = Registry.LocalMachine.OpenSubKey(@"HARDWARE\DEVICEMAP\SERIALCOMM"))
+            string[] allDevicesArray = null;
+            List<string> returnList = new List<string>();
+
+            while (returnSize == 0)
             {
-                if (serialKey != null)
+                unsafe
                 {
-                    string[] result = serialKey.GetValueNames();
-                    for (int i = 0; i < result.Length; i++)
+                    byte[] memBuff = new byte[maxSize];
+                    fixed (byte* memPtr = memBuff)
                     {
-                        // Replace the name in the array with its value.
-                        result[i] = (string)serialKey.GetValue(result[i]);
+                        returnSize = Interop.Kernel32.QueryDosDevice(null, (System.IntPtr)memPtr, maxSize);
+                        if (returnSize != 0)
+                        {
+                            allDevices = Marshal.PtrToStringAnsi((System.IntPtr)memPtr, (int)returnSize);
```

Comment:
This won't be necessary with the Unicode version of the API.

Reply:
Not sure if better way to do it, still refactoring this part to handle the buffer sizing loop like you suggested.
Not sure if better way to convert the unicode , but ended doiing 
`allDevice = System.Text.Encoding.Unicode.GetString(memBuff)`
for now, This also handled the fact QueryDosDeviceW() returns a double null terminating string, unlike QueryDosDevice that does not. Wasn't expecting that